pkg/discovery/grpcresolver/grpc_resolver.go

```diff
@@ -0,0 +1,167 @@
+// Copyright (c) 2019 The Jaeger Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://wwr.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package grpcresolver
+
+import (
+	"hash"
+	"hash/fnv"
+	"math/rand"
+	"sort"
+	"strconv"
+	"sync"
+	"time"
+
+	"go.uber.org/zap"
+	"google.golang.org/grpc/resolver"
+
+	"github.com/jaegertracing/jaeger/pkg/discovery"
+)
+
+// Resolver uses notifier to fetch list of available hosts
+type Resolver struct {
+	scheme            string
+	cc                resolver.ClientConn
+	notifier          discovery.Notifier
+	discoverer        discovery.Discoverer
+	logger            *zap.Logger
+	discoCh           chan []string // used to receive notifications
+	discoveryMinPeers int
+	salt              []byte
+
+	// used to block Close() until the watcher goroutine exits its loop
+	closing sync.WaitGroup
+}
+type hostScore struct {
+	address string
+	score   uint32
+}
+
+type hostScores []hostScore
+
+func (s hostScores) Len() int           { return len(s) }
+func (s hostScores) Swap(i, j int)      { s[i], s[j] = s[j], s[i] }
+func (s hostScores) Less(i, j int) bool { return s[i].score < s[j].score }
+
+// New intialize a new grpc resolver with notifier
+func New(
+	notifier discovery.Notifier,
+	discoverer discovery.Discoverer,
+	logger *zap.Logger,
+	discoveryMinPeers int,
+) *Resolver {
+	seed := time.Now().UnixNano()
+	random := rand.New(rand.NewSource(seed))
+	r := &Resolver{
+		notifier:          notifier,
+		discoverer:        discoverer,
+		discoCh:           make(chan []string, 100),
+		logger:            logger,
+		discoveryMinPeers: discoveryMinPeers,
+		salt:              []byte(strconv.FormatInt(random.Int63(), 10)), // random salt for rendezvousHash
+		scheme:            strconv.FormatInt(seed, 36),                   // make random scheme which will be used when registering
+	}
+	// TODO not sure if there's an equivalent way for grpc to maintain connection like what tchannel did?
```

Comment:
does this require follow-up?

Reply:
Please don't resolve comments if they are not addressed. I don't know what to do with this TODO.

Reply:
Tested it on my local setup. Roundrobin will include servers that's brought back automatically. Cheers.